### PR TITLE
Resolves #76

### LINKED ISSUE #76: Create a base class for agents
Our agents need to be given their action space, state space, and DRL model. It needs to have a way of interacting with the Environment and storing the rewards it receives for each interaction.

Suggestions:

- Class name: `AgentBase`.
- Class path: `urnai.agents`.
- Source: https://github.com/UFRN-URNAI/urnai-tools/blob/main/urnai/agents/sc2_agent.py .

### PR DESCRIPTION
**Pull request by CinquilCinquil**
#76

### CODE REVIEW

**Comment by CinquilCinquil** [comment]
I think the ActionSpace an agent receives could be replaced by an Strategy class (the class responsible for choosing the agent's next action), what do you think? @alvarofpp 

It should at least receive a Strategy class if im understanding their relationship correctly.

**Comment by CinquilCinquil** [comment]
It should at least receive a Strategy class if im understanding their relationship correctly.

**Comment by CinquilCinquil** [comment]
Hey, i tried rebasing to a newer version of v2 to make some adjustments in the code, but it appears that the old commits were duplicated when i commited the new changes. Any suggestions on how to fix this? or should i just ignore them?

**Comment by alvarofpp** [comment]
> Any suggestions on how to fix this? or should i just ignore them?

I recommend creating a new branch from `v2` and cherry-picking your commits about issue 76.

File: tests/units/agents/test_agent_base.py [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+
+class FakeAgent(AgentBase):
+    AgentBase.__abstractmethods__ = set()
+
+    def __init__(self, action_space,
+                 state_space : StateBase,
+                 model : ModelBase,
+                 reward_builder):
+        super().__init__(action_space, state_space, model, reward_builder)
+    ...
```
Comment (alvarofpp) [inline tests/units/agents/test_agent_base.py:18]: I don't think this `...` is necessary.

File: tests/units/agents/test_agent_base.py [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+
+class FakeState(StateBase):
+    StateBase.__abstractmethods__ = set()
+    ...
```
Comment (alvarofpp) [inline tests/units/agents/test_agent_base.py:22]: I don't think this `...` is necessary.

File: tests/units/agents/test_agent_base.py [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+
+class FakeModel(ModelBase):
+    ModelBase.__abstractmethods__ = set()
+
+    def ep_reset(self, ep):
+        return None
+    ...
```
Comment (alvarofpp) [inline tests/units/agents/test_agent_base.py:29]: I don't think this `...` is necessary.

File: tests/units/agents/test_agent_base.py [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+
+class FakeRewardBuilder:
+    def get_reward(self, obs, reward, done):
+        return None
+    def reset(self):
+        return None
+    ...
```
Comment (alvarofpp) [inline tests/units/agents/test_agent_base.py:36]: I don't think this `...` is necessary.

File: tests/units/agents/test_agent_base.py [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+
+class FakeActionSpace(ActionSpaceBase):
+    ActionSpaceBase.__abstractmethods__ = set()
+    ...
```
Comment (alvarofpp) [inline tests/units/agents/test_agent_base.py:40]: I don't think this `...` is necessary.

File: tests/units/agents/test_agent_base.py [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+
+    def test_reset(self):
+
+        # GIVEN
+        fake_action_space = FakeActionSpace()
+        fake_state_space = FakeState()
+        fake_model = FakeModel()
+        fake_reward_builder = FakeRewardBuilder()
+        fake_agent = FakeAgent(fake_action_space, fake_state_space,
+                               fake_model, fake_reward_builder)
```
Comment (alvarofpp) [inline tests/units/agents/test_agent_base.py:64]:
```suggestion
        fake_agent = FakeAgent(
            fake_action_space,
            fake_state_space,
            fake_model,
            fake_reward_builder
        )
```

File: tests/units/agents/test_agent_base.py [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+    ...
+
+class FakeActionSpace(ActionSpaceBase):
+    ActionSpaceBase.__abstractmethods__ = set()
+    ...
+
+class TestAgentBase(unittest.TestCase):
+
+    def test_abstract_methods(self):
+
```
Comment (alvarofpp) [inline tests/units/agents/test_agent_base.py:45]: It is not necessary to leave a blank line after the function signature.

File: tests/units/agents/test_agent_base.py [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+        # GIVEN
+        fake_state_space = FakeState()
+        fake_agent = FakeAgent(None, fake_state_space, None, None)
+
+        # WHEN
+        state_dim_return = fake_agent.state_dim
+
+        # THEN
+        self.assertEqual(state_dim_return, fake_state_space.dimension)
```
Comment (alvarofpp) [inline tests/units/agents/test_agent_base.py:121]:
```suggestion
    def test_state_dim(self):
        # GIVEN
        fake_state_space = FakeState()

        # WHEN
        fake_agent = FakeAgent(None, fake_state_space, None, None)

        # THEN
        self.assertEqual(fake_agent.state_dim, fake_state_space.dimension)
```

**Review by alvarofpp — CHANGES_REQUESTED** [review]

**Comment by CinquilCinquil** [comment]
To fix the duplicated commits i continued this issue on another branch (issue-76-2) on [this](https://github.com/UFRN-URNAI/urnai-tools/pull/94) pull request

**Pull request closed without merging**